File: compilers/python/runar_compiler/frontend/constant_fold.py
"""Constant folding pass for ANF IR.

Evaluates compile-time-known expressions and replaces them with ``load_const``
bindings.  Constants are propagated through the binding chain so downstream
operations can be folded too.

Direct port of ``compilers/rust/src/frontend/constant_fold.rs``.
"""

from __future__ import annotations

import json
from copy import deepcopy
from typing import Any

from runar_compiler.ir.types import (
    ANFBinding,
    ANFMethod,
    ANFProgram,
    ANFValue,
)


# ---------------------------------------------------------------------------
# Constant environment
# ---------------------------------------------------------------------------

# Constants are represented as (type_tag, value) tuples:
#   ("int", int_value)
#   ("bool", bool_value)
#   ("str", str_value)

ConstValue = tuple[str, Any]

ConstEnv = dict[str, ConstValue]


# ---------------------------------------------------------------------------
# Binary operation evaluation
# ---------------------------------------------------------------------------

def _eval_bin_op(op: str, left: ConstValue, right: ConstValue) -> ConstValue | None:
    # Arithmetic/bitwise/comparison on ints
    if left[0] == "int" and right[0] == "int":
        a: int = left[1]
        b: int = right[1]
        if op == "+":
            return ("int", a + b)
        if op == "-":
            return ("int", a - b)
        if op == "*":
            return ("int", a * b)
        if op == "/":
            if b == 0:
                return None
            # Truncated division (toward zero), matching JS BigInt semantics
            return ("int", int(a / b))
        if op == "%":
            if b == 0:
                return None
            # Remainder matching JS BigInt (sign follows dividend)
            return ("int", a - int(a / b) * b)
        if op == "===":
            return ("bool", a == b)
        if op == "!==":
            return ("bool", a != b)
        if op == "<":
            return ("bool", a < b)
        if op == ">":
            return ("bool", a > b)
        if op == "<=":
            return ("bool", a <= b)
        if op == ">=":
            return ("bool", a >= b)
        if op == "&":
            return ("int", a & b)
        if op == "|":
            return ("int", a | b)
        if op == "^":
            return ("int", a ^ b)
        if op == "<<":
            if a < 0:
                return None  # skip for negative left operand (BSV shifts are logical)
            if b < 0 or b > 128:
                return None
            return ("int", a << b)
        if op == ">>":
            if a < 0:
                return None  # skip for negative left operand (BSV shifts are logical)
            if b < 0 or b > 128:
                return None
            return ("int", a >> b)
        return None

    # Boolean operations
    if left[0] == "bool" and right[0] == "bool":
        a_b: bool = left[1]
        b_b: bool = right[1]
        if op == "&&":
            return ("bool", a_b and b_b)
        if op == "||":
            return ("bool", a_b or b_b)
        if op == "===":
            return ("bool", a_b == b_b)
        if op == "!==":
            return ("bool", a_b != b_b)
        return None

    # String (ByteString) operations
    if left[0] == "str" and right[0] == "str":
        a_s: str = left[1]
        b_s: str = right[1]
        if op == "+":
            if not _is_valid_hex(a_s) or not _is_valid_hex(b_s):
                return None
            return ("str", a_s + b_s)
        if op == "===":
            return ("bool", a_s == b_s)
        if op == "!==":
            return ("bool", a_s != b_s)
        return None

    # Cross-type equality
    if op == "===":
        return ("bool", False)
    if op == "!==":
        return ("bool", True)

    return None


def _is_valid_hex(s: str) -> bool:
    return all(c in "0123456789abcdefABCDEF" for c in s)


# ---------------------------------------------------------------------------
# Unary operation evaluation
# ---------------------------------------------------------------------------

def _eval_unary_op(op: str, operand: ConstValue) -> ConstValue | None:
    if operand[0] == "bool":
        if op == "!":
            return ("bool", not operand[1])
        return None
    if operand[0] == "int":
        n: int = operand[1]
        if op == "-":
            return ("int", -n)
        if op == "~":
            return ("int", ~n)
        if op == "!":
            return ("bool", n == 0)
        return None
    return None


# ---------------------------------------------------------------------------
# Builtin call evaluation (pure math functions only)
# ---------------------------------------------------------------------------

def _eval_builtin_call(func_name: str, args: list[ConstValue]) -> ConstValue | None:
    # Only fold pure math builtins with int arguments
    int_args: list[int] = []
    for a in args:
        if a[0] != "int":
            return None
        int_args.append(a[1])

    if func_name == "abs":
        if len(int_args) != 1:
            return None
        return ("int", abs(int_args[0]))

    if func_name == "min":
        if len(int_args) != 2:
            return None
        return ("int", min(int_args[0], int_args[1]))

    if func_name == "max":
        if len(int_args) != 2:
            return None
        return ("int", max(int_args[0], int_args[1]))

    if func_name == "safediv":
        if len(int_args) != 2 or int_args[1] == 0:
            return None
        return ("int", int(int_args[0] / int_args[1]))

    if func_name == "safemod":
        if len(int_args) != 2 or int_args[1] == 0:
            return None
        a, b = int_args[0], int_args[1]
        return ("int", a - int(a / b) * b)

    if func_name == "clamp":
        if len(int_args) != 3:
            return None
        val, lo, hi = int_args[0], int_args[1], int_args[2]
        return ("int", max(lo, min(val, hi)))

    if func_name == "sign":
        if len(int_args) != 1:
            return None
        n = int_args[0]
        if n > 0:
            return ("int", 1)
        elif n < 0:
            return ("int", -1)
        else:
            return ("int", 0)

    if func_name == "pow":
        if len(int_args) != 2:
            return None
        base, exp = int_args[0], int_args[1]
        if exp < 0 or exp > 256:
            return None
        result = 1
        for _ in range(exp):
            result *= base
        return ("int", result)

    if func_name == "mulDiv":
        if len(int_args) != 3 or int_args[2] == 0:
            return None
        tmp = int_args[0] * int_args[1]
        return ("int", int(tmp / int_args[2]))

    if func_name == "percentOf":
        if len(int_args) != 2:
            return None
        tmp = int_args[0] * int_args[1]
        return ("int", int(tmp / 10000))

    if func_name == "sqrt":
        if len(int_args) != 1:
            return None
        n = int_args[0]
        if n < 0:
            return None
        if n == 0:
            return ("int", 0)
        # Integer square root via Newton's method
        x = n
        y = (x + 1) // 2
        while y < x:
            x = y
            y = (x + n // x) // 2
        return ("int", x)

    if func_name == "gcd":
        if len(int_args) != 2:
            return None
        a, b = abs(int_args[0]), abs(int_args[1])
        while b != 0:
            a, b = b, a % b
        return ("int", a)

    if func_name == "divmod":
        if len(int_args) != 2 or int_args[1] == 0:
            return None
        return ("int", int(int_args[0] / int_args[1]))

    if func_name == "log2":
        if len(int_args) != 1:
            return None
        n = int_args[0]
        if n <= 0:
            return ("int", 0)
        return ("int", n.bit_length() - 1)

    if func_name == "bool":
        if len(int_args) != 1:
            return None
        return ("bool", int_args[0] != 0)

    return None


# ---------------------------------------------------------------------------
# ANF Value <-> ConstValue conversion
# ---------------------------------------------------------------------------

def _anf_value_to_const(value: ANFValue) -> ConstValue | None:
    """Extract a constant value from a load_const ANFValue."""
    if value.kind != "load_const":
        return None

    # Skip @ref: aliases -- they are binding references, not real constants
    if value.const_string is not None and value.const_string.startswith("@ref:"):
        return None

    # Check bool BEFORE int (bool is subclass of int in Python)
    if value.const_bool is not None:
        return ("bool", value.const_bool)
    if value.const_big_int is not None:
        return ("int", value.const_big_int)
    if value.const_int is not None:
        return ("int", value.const_int)
    if value.const_string is not None:
        return ("str", value.const_string)

    # Try to decode from raw_value
    raw = value.raw_value
    if raw is None:
        return None
    if isinstance(raw, bool):
        return ("bool", raw)
    if isinstance(raw, int):
        return ("int", raw)
    if isinstance(raw, float):
        return ("int", int(raw))
    if isinstance(raw, str):
        if raw.startswith("@ref:"):
            return None
        return ("str", raw)

    return None


def _const_to_anf_value(cv: ConstValue) -> ANFValue:
    """Convert a ConstValue to a load_const ANFValue."""
    tag, val = cv
    if tag == "int":
        return ANFValue(
            kind="load_const",
            raw_value=json.dumps(val),
            const_big_int=val,
            const_int=val,
        )
    if tag == "bool":
        return ANFValue(
            kind="load_const",
            raw_value=json.dumps(val),
            const_bool=val,
        )
    if tag == "str":
        return ANFValue(
            kind="load_const",
            raw_value=json.dumps(val),
            const_string=val,
        )
    # Fallback (shouldn't happen)
    return ANFValue(kind="load_const", raw_value=json.dumps(val))


# ---------------------------------------------------------------------------
# Fold bindings
# ---------------------------------------------------------------------------

def _fold_bindings(bindings: list[ANFBinding], env: ConstEnv) -> list[ANFBinding]:
    return [_fold_binding(b, env) for b in bindings]


def _fold_binding(binding: ANFBinding, env: ConstEnv) -> ANFBinding:
    folded_value = _fold_value(binding.value, env)

    # If the folded value is a load_const, register in the environment
    cv = _anf_value_to_const(folded_value)
    if cv is not None:
        env[binding.name] = cv

    return ANFBinding(name=binding.name, value=folded_value)


# ---------------------------------------------------------------------------
# Fold a single value
# ---------------------------------------------------------------------------

def _fold_value(value: ANFValue, env: ConstEnv) -> ANFValue:
    kind = value.kind

    if kind in ("load_const", "load_param", "load_prop"):
        return value

    if kind == "bin_op":
        left_const = env.get(value.left)
        right_const = env.get(value.right)
        if left_const is not None and right_const is not None:
            result = _eval_bin_op(value.op, left_const, right_const)
            if result is not None:
                return _const_to_anf_value(result)
        return value

    if kind == "unary_op":
        operand_const = env.get(value.operand)
        if operand_const is not None:
            result = _eval_unary_op(value.op, operand_const)
            if result is not None:
                return _const_to_anf_value(result)
        return value

    if kind == "call":
        if value.args is not None and all(a in env for a in value.args):
            const_args = [env[a] for a in value.args]
            folded = _eval_builtin_call(value.func, const_args)
            if folded is not None:
                return _const_to_anf_value(folded)
        return value

    if kind == "method_call":
        return value

    if kind == "if":
        cond_const = env.get(value.cond)
        if cond_const is not None and cond_const[0] == "bool":
            cond_val: bool = cond_const[1]
            if cond_val:
                then_env = dict(env)
                folded_then = _fold_bindings(value.then or [], then_env)
                # Merge constants from taken branch back into env
                for b in folded_then:
                    cv = _anf_value_to_const(b.value)
                    if cv is not None:
                        env[b.name] = cv
                return ANFValue(
                    kind="if",
                    cond=value.cond,
                    then=folded_then,
                    else_=[],
                )
            else:
                else_env = dict(env)
                folded_else = _fold_bindings(value.else_ or [], else_env)
                for b in folded_else:
                    cv = _anf_value_to_const(b.value)
                    if cv is not None:
                        env[b.name] = cv
                return ANFValue(
                    kind="if",
                    cond=value.cond,
                    then=[],
                    else_=folded_else,
                )
        else:
            # Condition not known -- fold both branches independently
            then_env = dict(env)
            else_env = dict(env)
            folded_then = _fold_bindings(value.then or [], then_env)
            folded_else = _fold_bindings(value.else_ or [], else_env)
            return ANFValue(
                kind="if",
                cond=value.cond,
                then=folded_then,
                else_=folded_else,
            )

    if kind == "loop":
        body_env = dict(env)
        folded_body = _fold_bindings(value.body or [], body_env)
        return ANFValue(
            kind="loop",
            count=value.count,
            body=folded_body,
            iter_var=value.iter_var,
        )

    # Terminal / side-effecting kinds pass through
    return value


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def _fold_method(method: ANFMethod) -> ANFMethod:
    env: ConstEnv = {}
    folded_body = _fold_bindings(method.body, env)
    return ANFMethod(
        name=method.name,
        params=list(method.params),
        body=folded_body,
        is_public=method.is_public,
    )


def fold_constants(program: ANFProgram) -> ANFProgram:
    """Apply constant folding to an ANF program.

    Evaluates compile-time-known expressions and replaces them with
    ``load_const`` bindings.  Does NOT run dead binding elimination --
    that is handled separately by the EC optimizer's DCE pass.
    """
    return ANFProgram(
        contract_name=program.contract_name,
        properties=list(program.properties),
        methods=[_fold_method(m) for m in program.methods],
    )

File: compilers/ruby/lib/runar_compiler/frontend/constant_fold.rb
# frozen_string_literal: true

# Constant folding pass for ANF IR.
#
# Evaluates compile-time-known expressions and replaces them with +load_const+
# bindings.  Constants are propagated through the binding chain so downstream
# operations can be folded too.
#
# Direct port of compilers/python/runar_compiler/frontend/constant_fold.py.

require "json"
require_relative "../ir/types"

module RunarCompiler
  module Frontend
    module ConstantFold
      # -----------------------------------------------------------------
      # Constant environment
      # -----------------------------------------------------------------
      #
      # Constants are represented as [type_tag, value] pairs:
      #   ["int", int_value]
      #   ["bool", bool_value]
      #   ["str", str_value]

      # -----------------------------------------------------------------
      # Binary operation evaluation
      # -----------------------------------------------------------------

      # @param op    [String]
      # @param left  [Array(String, Object)]
      # @param right [Array(String, Object)]
      # @return [Array(String, Object), nil]
      def self.eval_bin_op(op, left, right)
        # Arithmetic/bitwise/comparison on ints
        if left[0] == "int" && right[0] == "int"
          a = left[1]
          b = right[1]
          case op
          when "+"  then return ["int", a + b]
          when "-"  then return ["int", a - b]
          when "*"  then return ["int", a * b]
          when "/"
            return nil if b == 0
            # Truncated division (toward zero), matching JS BigInt semantics
            return ["int", _trunc_div(a, b)]
          when "%"
            return nil if b == 0
            # Remainder matching JS BigInt (sign follows dividend)
            return ["int", a - _trunc_div(a, b) * b]
          when "===" then return ["bool", a == b]
          when "!==" then return ["bool", a != b]
          when "<"   then return ["bool", a < b]
          when ">"   then return ["bool", a > b]
          when "<="  then return ["bool", a <= b]
          when ">="  then return ["bool", a >= b]
          when "&"   then return ["int", a & b]
          when "|"   then return ["int", a | b]
          when "^"   then return ["int", a ^ b]
          when "<<"
            return nil if a < 0   # skip for negative left operand (BSV shifts are logical)
            return nil if b < 0 || b > 128
            return ["int", a << b]
          when ">>"
            return nil if a < 0   # skip for negative left operand (BSV shifts are logical)
            return nil if b < 0 || b > 128
            return ["int", a >> b]
          else
            return nil
          end
        end

        # Boolean operations
        if left[0] == "bool" && right[0] == "bool"
          a_b = left[1]
          b_b = right[1]
          case op
          when "&&"  then return ["bool", a_b && b_b]
          when "||"  then return ["bool", a_b || b_b]
          when "===" then return ["bool", a_b == b_b]
          when "!==" then return ["bool", a_b != b_b]
          else
            return nil
          end
        end

        # String (ByteString) operations
        if left[0] == "str" && right[0] == "str"
          a_s = left[1]
          b_s = right[1]
          case op
          when "+"
            return nil unless valid_hex?(a_s) && valid_hex?(b_s)
            return ["str", a_s + b_s]
          when "===" then return ["bool", a_s == b_s]
          when "!==" then return ["bool", a_s != b_s]
          else
            return nil
          end
        end

        # Cross-type equality
        return ["bool", false] if op == "==="
        return ["bool", true]  if op == "!=="

        nil
      end
      private_class_method :eval_bin_op

      # Truncated division toward zero (matching JS BigInt semantics).
      def self._trunc_div(a, b)
        q = a.abs / b.abs
        q = -q if (a < 0) ^ (b < 0)
        q
      end
      private_class_method :_trunc_div

      def self.valid_hex?(s)
        /\A[0-9a-fA-F]*\z/.match?(s)
      end
      private_class_method :valid_hex?

      # -----------------------------------------------------------------
      # Unary operation evaluation
      # -----------------------------------------------------------------

      def self.eval_unary_op(op, operand)
        if operand[0] == "bool"
          return ["bool", !operand[1]] if op == "!"
          return nil
        end

        if operand[0] == "int"
          n = operand[1]
          case op
          when "-" then return ["int", -n]
          when "~" then return ["int", ~n]
          when "!" then return ["bool", n == 0]
          else
            return nil
          end
        end

        nil
      end
      private_class_method :eval_unary_op

      # -----------------------------------------------------------------
      # Builtin call evaluation (pure math functions only)
      # -----------------------------------------------------------------

      def self.eval_builtin_call(func_name, const_args)
        int_args = []
        const_args.each do |a|
          return nil unless a[0] == "int"
          int_args << a[1]
        end

        case func_name
        when "abs"
          return nil unless int_args.size == 1
          return ["int", int_args[0].abs]

        when "min"
          return nil unless int_args.size == 2
          return ["int", [int_args[0], int_args[1]].min]

        when "max"
          return nil unless int_args.size == 2
          return ["int", [int_args[0], int_args[1]].max]

        when "safediv"
          return nil unless int_args.size == 2 && int_args[1] != 0
          return ["int", _trunc_div(int_args[0], int_args[1])]

        when "safemod"
          return nil unless int_args.size == 2 && int_args[1] != 0
          a, b = int_args[0], int_args[1]
          return ["int", a - _trunc_div(a, b) * b]

        when "clamp"
          return nil unless int_args.size == 3
          val, lo, hi = int_args[0], int_args[1], int_args[2]
          return ["int", [lo, [val, hi].min].max]

        when "sign"
          return nil unless int_args.size == 1
          n = int_args[0]
          if n > 0
            return ["int", 1]
          elsif n < 0
            return ["int", -1]
          else
            return ["int", 0]
          end

        when "pow"
          return nil unless int_args.size == 2
          base, exp = int_args[0], int_args[1]
          return nil if exp < 0 || exp > 256
          result = 1
          exp.times { result *= base }
          return ["int", result]

        when "mulDiv"
          return nil unless int_args.size == 3 && int_args[2] != 0
          tmp = int_args[0] * int_args[1]
          return ["int", _trunc_div(tmp, int_args[2])]

        when "percentOf"
          return nil unless int_args.size == 2
          tmp = int_args[0] * int_args[1]
          return ["int", _trunc_div(tmp, 10_000)]

        when "sqrt"
          return nil unless int_args.size == 1
          n = int_args[0]
          return nil if n < 0
          return ["int", 0] if n == 0
          # Integer square root via Newton's method
          x = n
          y = (x + 1) / 2
          while y < x
            x = y
            y = (x + n / x) / 2
          end
          return ["int", x]

        when "gcd"
          return nil unless int_args.size == 2
          a, b = int_args[0].abs, int_args[1].abs
          a, b = b, a % b while b != 0
          return ["int", a]

        when "divmod"
          return nil unless int_args.size == 2 && int_args[1] != 0
          return ["int", _trunc_div(int_args[0], int_args[1])]

        when "log2"
          return nil unless int_args.size == 1
          n = int_args[0]
          return ["int", 0] if n <= 0
          return ["int", n.bit_length - 1]

        when "bool"
          return nil unless int_args.size == 1
          return ["bool", int_args[0] != 0]
        end

        nil
      end
      private_class_method :eval_builtin_call

      # -----------------------------------------------------------------
      # ANF Value <-> ConstValue conversion
      # -----------------------------------------------------------------

      # Extract a constant value from a load_const ANFValue.
      # @return [Array(String, Object), nil]
      def self.anf_value_to_const(value)
        return nil unless value.kind == "load_const"

        # Skip @ref: aliases -- they are binding references, not real constants
        if value.const_string && value.const_string.start_with?("@ref:")
          return nil
        end

        # Check bool BEFORE int
        unless value.const_bool.nil?
          return ["bool", value.const_bool]
        end
        unless value.const_big_int.nil?
          return ["int", value.const_big_int]
        end
        unless value.const_int.nil?
          return ["int", value.const_int]
        end
        unless value.const_string.nil?
          return ["str", value.const_string]
        end

        # Try to decode from raw_value
        raw = value.raw_value
        return nil if raw.nil?

        case raw
        when true, false
          return ["bool", raw]
        when Integer
          return ["int", raw]
        when Float
          return ["int", raw.to_i]
        when String
          return nil if raw.start_with?("@ref:")
          return ["str", raw]
        end

        nil
      end
      private_class_method :anf_value_to_const

      # Convert a ConstValue to a load_const ANFValue.
      def self.const_to_anf_value(cv)
        tag, val = cv
        v = IR::ANFValue.new(kind: "load_const")
        v.raw_value = JSON.generate(val)

        case tag
        when "int"
          v.const_big_int = val
          v.const_int = val
        when "bool"
          v.const_bool = val
        when "str"
          v.const_string = val
        end

        v
      end
      private_class_method :const_to_anf_value

      # -----------------------------------------------------------------
      # Fold bindings
      # -----------------------------------------------------------------

      def self.fold_bindings(bindings, env)
        bindings.map { |b| fold_binding(b, env) }
      end
      private_class_method :fold_bindings

      def self.fold_binding(binding, env)
        folded_value = fold_value(binding.value, env)

        # If the folded value is a load_const, register in the environment
        cv = anf_value_to_const(folded_value)
        env[binding.name] = cv unless cv.nil?

        IR::ANFBinding.new(
          name: binding.name,
          value: folded_value,
          source_loc: binding.source_loc
        )
      end
      private_class_method :fold_binding

      # -----------------------------------------------------------------
      # Fold a single value
      # -----------------------------------------------------------------

      def self.fold_value(value, env)
        kind = value.kind

        return value if kind == "load_const" || kind == "load_param" || kind == "load_prop"

        if kind == "bin_op"
          left_const = env[value.left]
          right_const = env[value.right]
          if left_const && right_const
            result = eval_bin_op(value.op, left_const, right_const)
            return const_to_anf_value(result) if result
          end
          return value
        end

        if kind == "unary_op"
          operand_const = env[value.operand]
          if operand_const
            result = eval_unary_op(value.op, operand_const)
            return const_to_anf_value(result) if result
          end
          return value
        end

        if kind == "call"
          if value.args && value.args.all? { |a| env.key?(a) }
            const_args = value.args.map { |a| env[a] }
            folded = eval_builtin_call(value.func, const_args)
            return const_to_anf_value(folded) if folded
          end
          return value
        end

        return value if kind == "method_call"

        if kind == "if"
          cond_const = env[value.cond]
          if cond_const && cond_const[0] == "bool"
            cond_val = cond_const[1]
            if cond_val
              then_env = env.dup
              folded_then = fold_bindings(value.then || [], then_env)
              # Merge constants from taken branch back into env
              folded_then.each do |b|
                cv = anf_value_to_const(b.value)
                env[b.name] = cv unless cv.nil?
              end
              new_v = IR::ANFValue.new(kind: "if")
              new_v.cond = value.cond
              new_v.then = folded_then
              new_v.else_ = []
              return new_v
            else
              else_env = env.dup
              folded_else = fold_bindings(value.else_ || [], else_env)
              folded_else.each do |b|
                cv = anf_value_to_const(b.value)
                env[b.name] = cv unless cv.nil?
              end
              new_v = IR::ANFValue.new(kind: "if")
              new_v.cond = value.cond
              new_v.then = []
              new_v.else_ = folded_else
              return new_v
            end
          else
            # Condition not known -- fold both branches independently
            then_env = env.dup
            else_env = env.dup
            folded_then = fold_bindings(value.then || [], then_env)
            folded_else = fold_bindings(value.else_ || [], else_env)
            new_v = IR::ANFValue.new(kind: "if")
            new_v.cond = value.cond
            new_v.then = folded_then
            new_v.else_ = folded_else
            return new_v
          end
        end

        if kind == "loop"
          body_env = env.dup
          folded_body = fold_bindings(value.body || [], body_env)
          new_v = IR::ANFValue.new(kind: "loop")
          new_v.count = value.count
          new_v.body = folded_body
          new_v.iter_var = value.iter_var
          return new_v
        end

        # Terminal / side-effecting kinds pass through
        value
      end
      private_class_method :fold_value

      # -----------------------------------------------------------------
      # Side-effect detection
      # -----------------------------------------------------------------

      SIDE_EFFECT_KINDS = %w[
        assert update_prop check_preimage deserialize_state
        add_output if loop call method_call
      ].freeze

      # Return true if this value kind has observable side effects.
      def self.has_side_effect(value)
        SIDE_EFFECT_KINDS.include?(value.kind)
      end

      # -----------------------------------------------------------------
      # Reference collection
      # -----------------------------------------------------------------

      # Walk an ANFValue and collect all binding name references.
      def self.collect_refs_from_value(value, used)
        if value.kind == "load_param"
          return
        end

        if value.kind == "load_const"
          if value.const_string && value.const_string.start_with?("@ref:")
            used.add(value.const_string[5..])
          end
          return
        end

        return if value.kind == "load_prop" || value.kind == "get_state_script"

        used.add(value.left)         if value.left
        used.add(value.right)        if value.right
        used.add(value.operand)      if value.operand
        used.add(value.cond)         if value.cond
        used.add(value.value_ref)    if value.value_ref
        used.add(value.object)       if value.object
        used.add(value.satoshis)     if value.satoshis
        used.add(value.preimage)     if value.preimage
        value.args&.each         { |a| used.add(a) }
        value.state_values&.each { |sv| used.add(sv) }
        value.then&.each  { |b| collect_refs_from_value(b.value, used) }
        value.else_&.each { |b| collect_refs_from_value(b.value, used) }
        value.body&.each  { |b| collect_refs_from_value(b.value, used) }
      end

      # -----------------------------------------------------------------
      # Public API
      # -----------------------------------------------------------------

      def self.fold_method(method)
        env = {}
        folded_body = fold_bindings(method.body, env)
        IR::ANFMethod.new(
          name: method.name,
          params: method.params.dup,
          body: folded_body,
          is_public: method.is_public
        )
      end
      private_class_method :fold_method

      # Apply constant folding to an ANF program.
      #
      # Evaluates compile-time-known expressions and replaces them with
      # +load_const+ bindings.  Does NOT run dead binding elimination --
      # that is handled separately by the EC optimizer's DCE pass.
      #
      # @param program [IR::ANFProgram]
      # @return [IR::ANFProgram]
      def self.fold_constants(program)
        IR::ANFProgram.new(
          contract_name: program.contract_name,
          properties: program.properties.dup,
          methods: program.methods.map { |m| fold_method(m) }
        )
      end
    end
  end
end
